app/maptools/recordingmaptool.cpp

```diff
@@ -275,15 +277,87 @@ void RecordingMapTool::setInitialGeometry( const QgsGeometry &newInitialGeometry
   if ( mInitialGeometry.equals( newInitialGeometry ) )
     return;
 
-  // We currently only support editing of points
-  if ( mLayer->geometryType() == QgsWkbTypes::PointGeometry )
-  {
-    QgsPoint point = QgsPoint( mInitialGeometry.asPoint() );
-    fixZ( point );
+  mInitialGeometry = newInitialGeometry;
 
-    mPoints.clear();
-    mPoints.push_back( point );
+  mPoints.clear();
+  for ( auto pointIt = mInitialGeometry.vertices_begin(); pointIt != mInitialGeometry.vertices_end(); ++pointIt )
+  {
+    mPoints.push_back( QgsPoint( *pointIt ) );
   }
 
   emit initialGeometryChanged( mInitialGeometry );
 }
+
+
+QgsGeometry RecordingMapTool::extractGeometryVertices( const QgsGeometry &geometry )
+{
+  QgsMultiPoint *multiPoint = new QgsMultiPoint();
+  QgsGeometry outputGeom( multiPoint );
+  for ( auto pointIt = geometry.vertices_begin(); pointIt != geometry.vertices_end(); ++pointIt )
+    multiPoint->addGeometry( ( *pointIt ).clone() );
+  return outputGeom;
+}
+
+QgsGeometry RecordingMapTool::extractMidSegmentVertices( const QgsGeometry &geometry )
+{
+  if ( geometry.type() == QgsWkbTypes::PointGeometry )
+  {
+    return QgsGeometry();
+  }
+
+  QgsMultiPoint *multiPoint = new QgsMultiPoint();
+  QgsGeometry outputGeom( multiPoint );
+  QgsPoint p;
+
+  const QVector< QgsLineString * > lines = QgsGeometryUtils::extractLineStrings( geometry.constGet() );
+  for ( QgsLineString *line : lines )
+  {
+    for ( int i = 0; i < line->numPoints() - 1; ++i )
+    {
+      p = QgsGeometryUtils::midpoint( line->pointN( i ), line->pointN( i + 1 ) );
+      multiPoint->addGeometry( p.clone() );
+    }
+
+    // if input geometry is a line we need to add virtual nodes and the beginning
+    // and at the end of the line. They will be used to extend line
+    if ( geometry.type() == QgsWkbTypes::LineGeometry )
+    {
+      p = QgsGeometryUtils::interpolatePointOnLine( line->pointN( 0 ), line->pointN( 1 ), -0.1 );
+      multiPoint->insertGeometry( p.clone(), 0 );
+      p = QgsGeometryUtils::interpolatePointOnLine( line->pointN( line->numPoints() - 2 ), line->pointN( line->numPoints() - 1 ), 1.1 );
+      multiPoint->addGeometry( p.clone() );
+    }
```

Comment:
What is the difference between this code and `RecordingMapTool::createHandles`?

Reply:
Here we add points at the beginning and end of the line, while `createHandles` creates lines which follow first or last segment direction